src/main/java/reactor/netty/transport/Transport.java

```diff
@@ -0,0 +1,318 @@
+	 * @return a new {@link Transport} reference
+	 */
+	public final <A> T attr(AttributeKey<A> key, @Nullable A value) {
+		Objects.requireNonNull(key, "key");
+		T dup = duplicate();
+		dup.configuration().attrs = TransportConfig.updateMap(configuration().attrs, key, value);
+		return dup;
+	}
+
+	/**
+	 * Provide a {@link ChannelGroup} - every active connected channel will be held in the
+	 * provided group.
+	 *
+	 * @param channelGroup a {@link ChannelGroup}
+	 * @return a new {@link Transport} reference
+	 */
+	public final T channelGroup(ChannelGroup channelGroup) {
+		Objects.requireNonNull(channelGroup, "channelGroup");
+		T dup = duplicate();
+		dup.configuration().channelGroup = channelGroup;
+		return dup;
+	}
+
+	/**
+	 * Return a {@link TransportConfig}.
+	 *
+	 * @return a {@link TransportConfig}
+	 */
+	public abstract C configuration();
+
+	/**
+	 * Set a new local address.
+	 *
+	 * @param localAddress a new local address
+	 * @return a new {@link Transport} reference
+	 */
+	public final T localAddress(SocketAddress localAddress) {
+		Objects.requireNonNull(localAddress, "localAddress");
+		T dup = duplicate();
+		dup.configuration().localAddress = localAddress;
+		return dup;
+	}
+
+	/**
+	 * Whether to enable metrics to be collected and registered in Micrometer's
+	 * {@link io.micrometer.core.instrument.Metrics#globalRegistry globalRegistry}.
+	 * Applications can separately register their own
+	 * {@link io.micrometer.core.instrument.config.MeterFilter filters}.
+	 * For example, to put an upper bound on the number of tags produced:
+	 * <pre class="code">
+	 * MeterFilter filter = ... ;
+	 * Metrics.globalRegistry.config().meterFilter(MeterFilter.maximumAllowableTags(prefix, 100, filter));
+	 * </pre>
+	 * <p>By default this is not enabled.
+	 *
+	 * @param enable true enables metrics collection; false disables it
+	 * @return a new {@link Transport} reference
+	 */
+	public final T metrics(boolean enable) {
+		if (enable) {
+			if (!Metrics.isInstrumentationAvailable()) {
+				throw new UnsupportedOperationException(
+						"To enable metrics, you must add the dependency `io.micrometer:micrometer-core`" +
+								" to the class path first");
+			}
+			T dup = duplicate();
+			dup.configuration().metricsRecorder = configuration().defaultMetricsRecorder();
+			return dup;
+		}
+		else if (configuration().metricsRecorder != null) {
+			T dup = duplicate();
+			dup.configuration().metricsRecorder = null;
+			return dup;
+		}
+		else {
+			@SuppressWarnings("unchecked")
+			T dup = (T) this;
+			return dup;
+		}
+	}
+
+	/**
+	 * Specifies whether the metrics are enabled on the {@link Transport}.
+	 * All generated metrics are provided to the specified recorder.
+	 *
+	 * @param enable if true enables the metrics on the {@link Transport}.
+	 * @param recorder the {@link ChannelMetricsRecorder}
+	 * @return a new {@link Transport} reference
+	 */
+	public final T metrics(boolean enable, ChannelMetricsRecorder recorder) {
```

Comment:
The idea is that you might have system property or env variable that controls whether metrics are enabled or not.

`metrics(boolean)` - enables/disables the default implementation with Micrometer while
`metrics(boolean, ChannelMetricsRecorder)`- enables/disables the metrics and uses a custom ChannelMetricsRecorder not the default with Micrometer

Reply:
but wouldn't it be weird to have a (nonNull) `recorder` parameter along with a `boolean false` value for `enable`? I kinda see the usage, but I fear it might get confusing. For instance, say I do `.metrics(false, new MyFancyMetricsRecorder())`. Then later on, I call `.metrics(true)` with the sense that it will activate the fancy metrics recorder I prepared earlier. But instead, it activates the micrometer one (because the `MyFancyMetricsRecorder` was never saved).

even if it makes it a tiny bit more complicated to use with a sysprop, we might instead go this way:
```java
//activate metrics, use the provided recorder
metricsRecordedWith(MetricsRecorder recorder);
//activate metrics, use the default recorder
metricsRecordedWithMicrometer();
//deactivate metrics
metricsDisabled();
```
wdyt @violetagg @rstoyanchev ?

Reply:
Javadoc clearly indicates the usage and the difference

```
	/**
	 * Whether to enable metrics to be collected and registered in Micrometer's
	 * {@link io.micrometer.core.instrument.Metrics#globalRegistry globalRegistry}.
	 * Applications can separately register their own
	 * {@link io.micrometer.core.instrument.config.MeterFilter filters}.
	 * For example, to put an upper bound on the number of tags produced:
	 * <pre class="code">
	 * MeterFilter filter = ... ;
	 * Metrics.globalRegistry.config().meterFilter(MeterFilter.maximumAllowableTags(prefix, 100, filter));
	 * </pre>
	 * <p>By default this is not enabled.
	 *
	 * @param enable true enables metrics collection; false disables it
	 * @return a new {@link Transport} reference
	 */
	public final T metrics(boolean enable) {
```

```
	/**
	 * Specifies whether the metrics are enabled on the {@link Transport}.
	 * All generated metrics are provided to the specified recorder.
	 *
	 * @param enable if true enables the metrics on the {@link Transport}.
	 * @param recorder the {@link ChannelMetricsRecorder}
	 * @return a new {@link Transport} reference
	 */
	public final T metrics(boolean enable, ChannelMetricsRecorder recorder) {
```

Till now there were no complains about the methods usage.